modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientTable.java

```diff
@@ -839,25 +868,202 @@ <E> CompletableFuture<List<E>> split(
+        }
+
+        CompletableFutures.allOf(res).handle((ignored, err) -> {
+            // TODO remove copy paste
+            List<CompletableFuture<Void>> waitCommitFuts = List.of();
+            if (!txns.isEmpty()) {
+                boolean allRetryableExceptions = true;
+
+                for (int i = 0; i < res.size(); i++) {
+                    CompletableFuture<?> fut0 = res.get(i);
+                    if (fut0.isCompletedExceptionally()) {
+                        try {
+                            fut0.join();
+                        } catch (CompletionException e) {
+                            allRetryableExceptions = ExceptionUtils.matchAny(unwrapCause(e), ACQUIRE_LOCK_ERR);
                         }
+                    }
+                    txns.get(i).rollbackAsync();
+                }
 
-                        return in;
-                    });
+                if (err != null) {
+                    // Check if we can retry.
+                    long nowRelative = System.nanoTime();
+                    if (allRetryableExceptions && nowRelative - startTs[0] < DEFAULT_IMPLICIT_GET_ALL_TIMEOUT_NANOS) {
+                        startTs[0] = nowRelative;
+                        txns.clear(); // This collection is re-filled on next map attempt.
+
+                        mapAndRetry(mapFun, keys, txns, mapped, startTs, resFut, log);
+
+                        return null;
+                    }
+
+                    resFut.completeExceptionally(err);
+
+                    return null;
+                }
+
+                waitCommitFuts = unlockFragments(txns, log);
+            } else {
+                if (err != null) {
+                    resFut.completeExceptionally(err);
+
+                    return null;
+                }
+            }
+
+            var in = new ArrayList<E>(Collections.nCopies(keys.size(), null));
+
+            for (int i = 0; i < res.size(); i++) {
+                CompletableFuture<List<E>> f = res.get(i);
+                reduceWithKeepOrder(in, f.getNow(null), batches.get(i).originalIndices);
+            }
+
+            if (waitCommitFuts.isEmpty()) {
+                resFut.complete(in);
+            } else {
+                CompletableFutures.allOf(waitCommitFuts).whenComplete((r, e) -> {
+                    // Ignore errors.
+                    resFut.complete(in);
                 });
+            }
+
+            return null;
+        });
+    }
+
+    @NotNull
+    private static List<CompletableFuture<Void>> unlockFragments(List<Transaction> txns, IgniteLogger log) {
+        List<CompletableFuture<Void>> waitCommitFuts = new ArrayList<>();
+
+        for (Transaction txn : txns) {
+            // ClientTransaction tx0 = (ClientTransaction) txn; TODO FIXME investigate error handling
+            ClientLazyTransaction tx0 = (ClientLazyTransaction) txn;
+            CompletableFuture<Void> fut = tx0.commitAsync().whenComplete((r, e) -> {
```

Comment:
Why do you commit transactions here if you have already rolled them back in the code above?